jre_emul/Classes/dalvik/system/VMStack.java

```diff
@@ -14,11 +16,101 @@
 
 package dalvik.system;
 
+import android.compat.annotation.UnsupportedAppUsage;
+
+import dalvik.annotation.optimization.FastNative;
+
 /**
- * Stubbed version of Android's VMStack.
+ * Provides a limited interface to the Dalvik VM stack. This class is mostly
+ * used for implementing security checks.
+ *
+ * @hide
  */
+@libcore.api.CorePlatformApi
 public final class VMStack {
-  public static ClassLoader getCallingClassLoader() {
-    return null;
+
+  private VMStack() {
   }
+
+  /**
+   * Returns the defining class loader of the caller's caller.
+   *
+   * @return the requested class loader, or {@code null} if this is the
+   *         bootstrap class loader.
+   * @deprecated Use {@code ClassLoader.getClassLoader(sun.reflect.Reflection.getCallerClass())}.
+   *         Note that that can return {@link BootClassLoader} on Android where the RI
+   *         would have returned null.
+   */
+  @UnsupportedAppUsage
+  @FastNative
+  @Deprecated
+  native public static ClassLoader getCallingClassLoader();
+
+  /**
+   * Returns the class of the caller's caller.
+   *
+   * @return the requested class, or {@code null}.
+   * @deprecated Use {@link sun.reflect.Reflection#getCallerClass()}.
+   */
+  @Deprecated
+  public static Class<?> getStackClass1() {
+    return getStackClass2();
+  }
+
+  /**
+   * Returns the class of the caller's caller's caller.
+   *
+   * @return the requested class, or {@code null}.
+   */
+  @UnsupportedAppUsage
+  @FastNative
+  native public static Class<?> getStackClass2();
```

Comment:
This can be implemented using something like (not tested):
```
StackTraceElement[] stackTrace = getThreadStackTrace(Thread.currentThread());
return Class.forName(stackTrace[3].getClassName());
```
Embedded native code may be faster.